File: python/20_Valid_Parentheses.py
#!/usr/bin/python
import sys
import os

class Solution(object):
    def isValid(self, s):
        """
        :type s: str
        :rtype: bool
        """
        stk = []
        for i in range(len(s)):
        	if s[i] == '(' or s[i] == '[' or s[i] == '{':
        		stk.append(s[i])
        	else:
        		if stk == [] or self.unmatch(stk.pop(), s[i]):
        			return False
        return stk == []

    def unmatch(self, a, b):
    	if a == '(':
    		return b != ')'
    	elif a == '[':
    		return b != ']'
    	elif a == '{':
    		return b != '}'
    	else:
    		return False



def main():
    s = Solution()
    print s.isValid("[")

if __name__ == "__main__":
    main()

File: c++/20_Valid_Parentheses.cpp
#include <iostream>
#include <vector>
#include <map>
#include <cmath>
#include <stack>
#include <string>

using namespace std;

class Solution {
public:
    bool isValid(string s) {
        stack<char> stk;

        for (int i = 0; i < s.size(); i++) {
        	if (s[i] == '(' || s[i] == '[' || s[i] == '{')
        		stk.push(s[i]);
        	else {
        		if (stk.empty() || unmatch(stk.top(), s[i]))
        			return false;
        		stk.pop();
        	}
        }
        return stk.empty();
    }
    bool unmatch(char a, char b) {
    	if (a == '(')
    		return (b != ')');
    	else if (a == '[')
    		return (b != ']');
    	else if (a == '{')
    		return (b != '}');
    	else
    		return false;
    }
};


int main() {
    Solution s;
    
}
